### DISCUSSION TITLE
Lesson 12: Codehawks review - Gas Optimisation

### DISCUSSION THREAD

**Discussion by sunnyStefi** [Cyfrin/foundry-full-course-cu#999] [Q&A]
I'm going through the [codehawks audit](https://www.codehawks.com/report/cljx3b9390009liqwuedkn0m0#G-01) for the Stablecoin contract.
Inside the Section Gas Optimisation, I cannot replicate the point G01 and G03 when i run` forge test --gas-report`: instead of cheaper functions I get more expensive one.

I'm testing with a simple contract:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

contract GasOptimisation {
    uint256 public a;
    uint256 public b;
    uint256 constant MAXUINT256 = 115792089237316195423570985008687907853269984665640564039457584007913129639935;

    //44562 gas --> CHEAPER, not EXPENSIVE
    function shortHandNotation() public {
        a = 1;
        b = 2;
        a += b;
    }

    //44584 gas --> EXPENSIVE, not CHEAPER
    function normalNotation() public {
        a = 1;
        b = 2;
        a = a + b;
    }

    //206 gas --> CHEAPER, not EXPENSIVE
    function callingTypeMaxDirectly() public returns (uint256 test) {
        test = type(uint256).max;
    }

    //250 gas --> EXPENSIVE, not CHEAPER
    function callingConstantTypeMax() public returns (uint256 test) {
        test = MAXUINT256;
    }
}
```

And testing those inside a test contract in the simplest way possible:

```solidity
pragma solidity ^0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {GasOptimisation} from "../src/GasOptimisation.sol";

contract GasOptimisationTest is Test {
    GasOptimisation public gasOptimisator;

    function setUp() external {
        gasOptimisator = new GasOptimisation();
    }

    function test_normalNotation() public {
        gasOptimisator.normalNotation();
    }
...
```

Why are the results of `forge snapshot` different from `forge test --gas-report`?

**Comment by 0xusmanf** [comment]
> Inside the Section Gas Optimisation, I cannot replicate the point G01 and G03 when i run forge test --gas-report: instead of cheaper functions I get more expensive one.

Yup, you are right, tried it with your provided code and the results were not as expected. Maybe it has something to do with the solidity version, but I am not sure.
```shell
| src/GasOptimisation.sol:GasOptimisation contract |                 |       |        |       |         |
|--------------------------------------------------|-----------------|-------|--------|-------|---------|
| Deployment Cost                                  | Deployment Size |       |        |       |         |
| 59511                                            | 329             |       |        |       |         |
| Function Name                                    | min             | avg   | median | max   | # calls |
| callingConstantTypeMax                           | 271             | 271   | 271    | 271   | 1       |
| callingTypeMaxDirectly                           | 249             | 249   | 249    | 249   | 1       |
| normalNotation                                   | 44584           | 44584 | 44584  | 44584 | 1       |
| shortHandNotation                                | 44562           | 44562 | 44562  | 44562 | 1       |
```

> Why are the results of `forge snapshot` different from `forge test --gas-report`?

#### Gas reports and gas snapshots differ in some ways:
* Gas reports use tracing to figure out gas costs for individual contract calls.
This gives more granular insight, at the cost of speed.
* Gas snapshots have more built-in tools, such as diffs and exporting the results to a file.
Snapshots are not as granular as gas reports, but they are faster to generate.

[SOURCE](https://book.getfoundry.sh/forge/gas-tracking)

**Reply by sunnyStefi** [reply]
Thank you for taking some time to look at it. From the Foundry Book they don't explain where these values are coming from and how are they calculated

**Reply by 0xusmanf** [reply]
This is what I was able to find regarding `forge snapshot` and `forge test --gas-report` producing different results. Would love to know more, please do share if you find any other resource detailing the reason.

**Reply by 0xusmanf** [reply]
Another explanation I got [in a discord chat](https://discord.com/channels/1128482318634471526/1135651030642729060/1162006393809809548)

> can find a link later but basically snapshot records the gas consumed by the full test functions, --gas-report records the gas consumed by the contract's functions themselves

Meaning `--gas-report` will always give you a static result regardless of how you implement your tests but `snapshot` on the other hand will give you results based on how you implement your tests.
